### ISSUE TITLE
use redis to cache clusterIP of thing servers to enable point-to-point messaging to things and/or improve pubsub scaling

### ISSUE THREAD

**Issue by sking2600** [absmach/magistrala#456]
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.
No

2. Describe the feature you are requesting, as well as the possible use case(s) for it.

the OCF specification for device<=>cloud communication requires that the device establish a long-lived connection to the cloud. This makes the CoAP servers stateful and raises the question, "how do I route messages to the device?" which must be solved by having some mapping between the device UUID and the k8s clusterIP of the CoAP server which is serving that device. The current architecture of mainflux utilizes NATS pubsub to solve this problem (the mapping is between channels named after the device UUID and subscribers to that channel) however that requires 1 channel per device and NATS was not designed to scale easily to 10's of millions of topics. 

I propose that we utilize redis to map the deviceUUID and coap server clusterIP to enable clients (within the k8s cluster) to communicate with the coap server directly instead of relying on a message broker. I'm assuming that the latency is lower with redis or, at the very least, the time complexity of looking up a UUID:clusterIP key-value pair is less than publishing a request (https://redis.io/commands/publish). It may also be preferred to use redis over NATS for pubsub due to potentially easier scaling to 10's of millions of topics (aka use a "master" redis node to tell you which "slave" node has the device that you want to SUBSCRIBE to)

before: external client -> HTTP server -> NATS broker -> coap server -> device
after: external client -> HTTP server (looks up clusterIP based on target UUID) -> coap server -> device 

rough draft design doc:
https://docs.google.com/document/d/11QHlPieNdlUoTmAniUio7MXiun3ykh6_3u8eZjo25OM/edit?usp=sharing

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
I can't currently imagine a more efficient/scalable approach to the problem of supporting long-lived CoAP+TCP connections

comments: 
-the point-to-point messaging bit can be implemented without replacing the current usage of NATS for devices publishing data.
-I'm assuming there would be interest in supporting OCF compliant CoAP devices (this would mitigate the need for UDP/DTLS support and thus mitigate #428)
-you'd need to scale horizontally more so than vertically with the redis instances due to the time complexity of the PUBLISH command (https://redis.io/commands/publish)
-the issue with pubsub scaling and message routing are separate issues. I just happen to be proposing that they both be solved with redis

**Comment by drasko** [comment]
I think this should be solved though a scalable and clusterable CoAP server. Current Go CoAP implementation is not clusterable (does not exchange internal state via some usual mechanisms - for example RAFT).

Closing this due inactivity.
